### PR TITLE
[MLIR][Mesh] Add sharding propagation pass

### PR DESCRIPTION
**Pull request by yaochengji**
- add `ShardingInterface` and the methods' default implementation
- add `ShardingInterface` implementation for element-wise and matmul ops in TOSA dialect
- add sharding propagation pass
- add Mesh_IteratorType enum

Small changes in non-mesh folder:
- add `getMultiDimMapWithTargets` method in AffineMap.h

### CODE REVIEW

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:38]: This should move directly behind an API, the `getAttr` method of Operation are intended to be deprecated (to use `getDiscardableAttr` /`getInherentAttr` instead) and implementer of this interface could put these in the Properties storage, and so the best way is to expose all these accessors through the `OpInterface`.

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:129]: Why this copy?

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:141]: Same question: why the copy here?

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:120]: I don't quite get what the DenseSet is doing here?

**Review by joker-eph — COMMENTED** [review]
Thanks for the PR!

A few early comments after a quick scan.

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:129]: Because new operations (`mesh.shard` ops) will be added during propagation.

File: mlir/lib/IR/Builders.cpp
Comment (yaochengji) [reply mlir/lib/IR/Builders.cpp]: Done.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp]: Added `getShardingOptionFromAttr` and `setShardingOptionAttr` in extraClassDeclaration

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp]: It shouldn't be there, my bad.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:60]: Good catch, it's not needed.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:24]: Done.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:45]: Done.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:282]:
```suggestion
  for (size_t i = 0; i < shardingOption.shardingArray.size(); ++i) {
    if (i == loopIdx) continue;
    for (int32_t axis : meshAxes) {
      if (std::find(shardingOption.shardingArray[i].begin(),
                    shardingOption.shardingArray[i].end(),
                    axis) != shardingOption.shardingArray[i].end()) {
        if (ignoreIfConflicted) {
          return success();
        } else {
          return op->emitOpError()
                 << "sharding option confilicts because of mesh axis " << axis
                 << " duplicates";
        }
      }
    }
```

(nit: early continue)

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (joker-eph) [reply mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:116]: I only see `setShardingOptionAttr(builder, *shardingOption);` and `shardingOp.addShardingAnnotations`, where are these added?
Also, wouldn't `for (Operation &op : llvm::make_early_inc(block.getOperations())) {` still work? Assuming the `shard` ops are added immediately after the current operation (otherwise they can be also just skipped ...)

**Review by joker-eph — COMMENTED** [review]
Some more superficial comment, I haven't dug deeply in the logic yet. Would be good to also have another pair of eyes.

Also what about control flow operations?

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td]: Sorry it seems too many typos in this PR. I have proofread the entire PR for spelling errors, and I hope there will be fewer typos in the future.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp]: Change to `return failure()` if `val.hasOneUse()`.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td]: I removed the logic of adding `ShardingOption` as an attribute first. Because in my second thought, this information might be more suitable in an analysis. 

BTW, the reason they were in `extraClassDeclaration` not `InterfaceMethod` is because they are not supposed to be customized for different operations.

**Comment by yaochengji** [comment]
> I only see setShardingOptionAttr(builder, *shardingOption); and shardingOp.addShardingAnnotations, where are these added?

They are defined in `ShardingInterface.cpp`.

> Also, wouldn't for (Operation &op : llvm::make_early_inc(block.getOperations())) { still work? Assuming the shard ops are added immediately after the current operation (otherwise they can be also just skipped ...)

Yeah, ` llvm::make_early_inc_range` is better.

**Comment by yaochengji** [comment]
> Some more superficial comment, I haven't dug deeply in the logic yet. Would be good to also have another pair of eyes.

@sogartar could you also review this PR if you have time?


> Also what about control flow operations?

The sharding propagation pass is currently in an early stage. I believe that more comprehensive support needs to be added in the future, including but not limited to control flow support. (Maybe I should add more TODOs in the code)

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:93]: Clang-format seems to work fine here for me?
```
    LLVM_DEBUG(
        DBGS() << "print all the ops' iterator types and indexing maps in the "
                  "block.\n";
        for (Operation &op
             : block.getOperations()) {
          if (auto shardingOp = llvm::dyn_cast<ShardingInterface>(&op))
            shardingOp.printLoopTypesAndIndexingMaps(llvm::dbgs());
        });
```

(it fixes your indentation issue line 90 where the stmt inside the `if` isn't indented right now)

Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp]: Done.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (sogartar) [inline mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:63]: Isn't this method more related to the auto-sharding algorithm than it is to how an operation is sharded. It is very particular to how sharding propagation functions.
Shouldn't there be a more general method that enumerates what are the possible ways the operation can be shardings? Then the sharding propagation could chose based on the existing sharding for some of the operands/results.
`ShardingOption` seems to describe only one way an operation is sharded. Shouldn't it be like a collection? Maybe not explicitly a collection as its size may be very large.

**Comment by sogartar** [comment]
> @sogartar could you also review this PR if you have time?

I left some comments.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (sogartar) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:63]: I think this method should return the list of possible op shardings that are compatible (do not require resharding) with the already annotated operands/results.
Maybe we can have a more general form where you are able to query for sharding options where operands/results shardings are passed as method arguments rather than being read from the IR.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:71]: I am concerned that some special operations, such as control flow operations, will require custom implementations. I have not yet fully considered what these implementations would look like.

File: mlir/lib/Dialect/Utils/IndexingUtils.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Utils/IndexingUtils.cpp]: Done.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:117]: It is possible that the code will execute. I made some minor changes to the code to improve readability.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp]: No. You can try `#map0 = affine_map<(d0, d1) -> (3 * d0, d1)>` and will find it's a valid affine map.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:426]: The cast won't fail as long as the result's indexing map satisfies the `isProjectedPermutation` resctriction.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:159]: I think tensor<f32> could also be a scalar.

Another important reason is that `RankedTensorType` supports an additional encoding attribute which I will use later.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:153]: Yeah, in case other users want to verify their implementation of `getLoopIteratorTypes` and `getIndexingMaps`

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:183]: Conv should be fine as it only requires the op results to be projected permutations. In your example, the indexing map of the result is ` affine_map<(d0, d1, d2, d3, d4, d5, d6)[s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10] -> (d0, d1, d2, d3)>`, and it's a projected permutation.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp]: Done.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp]: Done.

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp]: Done.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:63]: Because by design, the sharding propagation pass only deduces the "simplest " `ShardingOption` for each operation, not the best one. Therefore, frontend users can infer as much as possible what the resulting IR is with partial sharding annotations.

Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:63]: I think a new `getPossibleShardingOptions` method could also be useful in some use cases. E.g. we want to deduce the sharding algorithm for the whole graph progressively. 

And I think passing operands/results shardings as arguments instead of reading from the IR is a great idea. Thus we don't need to actually insert operations into the IR. It could be useful in analyzing sharding algorithm.

Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:63]: I will modify the `getShardingOption` method to passing operands/results shardings as arguments in this PR. It should be ready tomorrow.

But for the `getPossibleShardingOptions` change, it might require more consideration. For example, returning only the possible sharding options is not enough, they should also include the communication costs. So I think a  the current `getShardingOption` which only returns one `ShardingOption` is fine at current stage. At least the expected sharding strategy could be get as long as proper `mesh.shard` ops are inserted.

**Comment by yaochengji** [comment]
> I left some comments.

@sogartar,  Thank you for your careful review. I have responded to all of your comments.

**Comment by sogartar** [comment]
@yaochengji, I marked most of the discussions as resolved.

File: mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:63]: @sogartar the `getShardingOption` method is changed to passing operands/results shardings as arguments version.

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (sogartar) [inline mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:41]: Could you mention in the comments what is the ordering?

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (sogartar) [inline mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:134]: Do we here try to pick first the sharding option that corresponds to the most specific operand shardings?
Why do we treat operands and results differently?

**Review by sogartar — APPROVED** [review]

File: mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:158]: I'm afraid not. For example, the `argmax` op returns a scalar index, but it's not replicated for the distributed version of `argmax`.

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp:42]: Done.

File: mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/Transforms/ShardingPropagation.cpp]: > Do we here try to pick first the sharding option that corresponds to the most specific operand shardings?

Yes we do.

> Why do we treat operands and results differently?

They should be treated the same. I'm sorry I just forgot about the results' logic. It is now modified.

**Comment by yaochengji** [comment]
Hi @joker-eph , do you think this PR could be merged? As @sogartar already approved this.

File: mlir/lib/IR/Builders.cpp
Comment (joker-eph) [inline mlir/lib/IR/Builders.cpp:326]: This isn't used?

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (joker-eph) [inline mlir/include/mlir/Dialect/Utils/IndexingUtils.h:249]: Seems unused?

File: mlir/lib/IR/AffineMap.cpp
Comment (joker-eph) [inline mlir/lib/IR/AffineMap.cpp:238]: @Groverkss wanna double check this? (or @ftynse maybe?)

**Review by joker-eph — COMMENTED** [review]
Looks fine, just minor things to cleanup and I can merge this for you.

File: mlir/lib/IR/Builders.cpp
Comment (yaochengji) [reply mlir/lib/IR/Builders.cpp]: It isn't used currently. Previous this was used by the sharding option attribute logic. The PR description has also been modified to align this change.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (yaochengji) [reply mlir/include/mlir/Dialect/Utils/IndexingUtils.h]: done.

File: mlir/lib/IR/AffineMap.cpp
Comment (Groverkss) [inline mlir/lib/IR/AffineMap.cpp:242]: You don't need to pass numDims here. You can just infer it. See getProjectedMap

File: mlir/lib/IR/AffineMap.cpp
Comment (Groverkss) [inline mlir/lib/IR/AffineMap.cpp:245]: You can do this the same way getPermutationMap does. You could even make getPermutationMap call to this function for some clean up.

Comment (yaochengji) [reply mlir/lib/IR/AffineMap.cpp]: Thanks for your suggestion. I have modified the code accordingly.

**Comment by joker-eph** [comment]
I tweaked the description and landed it, thanks!

**Comment by yaochengji** [comment]
Some tests failed. It seems that `MLIRFuncDialect` and `MLIRPass` are missing in `MLIRMeshTransforms`. How can I fix that as this PR is already merged. @joker-eph

**Comment by joker-eph** [comment]
Seems like the CI does not merge before testing?
I'll revert and you can cherry-pick the original commit and open a new PR with the fixes.

**Comment by joker-eph** [comment]
See also the fix in bda763aea0b854178c01eac9f309042d9aaa823b to add to your next PR

**Comment by yaochengji** [comment]
Thanks @joker-eph, the [new PR](https://github.com/llvm/llvm-project/pull/71261) is created, and the fix in https://github.com/llvm/llvm-project/commit/bda763aea0b854178c01eac9f309042d9aaa823b is added.

**Pull request merged**
